src/xtgeo/io/tsurf/_tsurf_reader.py

```diff
@@ -0,0 +1,799 @@
+        [x,y,z] coordinates and 'triangles' is a list of vertex indices
+        (starting with 1)
+
+    Raises:
+        ValueError: If TFACE section contains invalid lines
+    """
+
+    vertices: list[list[float]] = []
+    triangles: list[list[int]] = []
+
+    err_msg_vrtx = (
+        f"\nIn file {filepath_errmsg}:\n"
+        "Invalid 'VRTX' line in 'TFACE' section.\n"
+        "Expected format: 'VRTX id x y z [attributes]'\n"
+        "where:\n"
+        "  - 'id' starts with '1' and increments by 1 for each vertex\n"
+        "  - 'x', 'y', and 'z' are floating-point numbers\n"
+        "  - the only currently supported attribute is 'CNXYZ'."
+    )
+
+    err_msg_trgl = (
+        f"\nIn file {filepath_errmsg}:\n"
+        "Invalid 'TRGL' line in 'TFACE' section.\n"
+        "Expected format: 'TRGL vertex_id1 vertex_id2 vertex_id3'\n"
+        "where:\n"
+        "  - each vertex id must be an integer\n"
+        "  - each vertex id must be in the range "
+        "{{1, number of vertices}} (1-based indexing)\n"
+    )
+
+    # Keep track of expected vertex numbering, 1-based numbering
+    vrtx_no = 0
+
+    # Is 'END' statement present?
+    end_is_present = False
+
+    for line in _read_line(stream):
+        if line[0] == "VRTX":
+            if len(line) != 6 or line[1] != str(vrtx_no + 1) or line[5] != "CNXYZ":
+                err_msg_vrtx += f"Failing line: '{line}'"
+                raise ValueError(err_msg_vrtx)
+
+            try:
+                vertices.append([float(line[2]), float(line[3]), float(line[4])])
+            except ValueError:
+                err_msg_vrtx += f"Failing line: '{line}'"
+                raise ValueError(err_msg_vrtx)
+
+            vrtx_no += 1
+
+        elif line[0] == "TRGL":
+            if len(line) != 4:
+                err_msg_trgl += f"Failing line: '{line}'"
+                raise ValueError(err_msg_trgl)
+
+            try:
+                triangles.append([int(line[1]), int(line[2]), int(line[3])])
+            except ValueError:
+                err_msg_trgl += f"Failing line: '{line}'"
+                raise ValueError(err_msg_trgl)
+
+        elif line[0] == "END":
+            end_is_present = True
+            break
+
+        else:
+            raise ValueError(
+                f"\nIn file {filepath_errmsg}:\n"
+                "Invalid line in 'TFACE' section with triangulated data.\n"
+                "Expect lines to start with 'VRTX', 'TRGL', or 'END'.\n"
+                f"Failing line: '{line}'",
+            )
+
+    if not end_is_present:
+        raise ValueError(
+            f"\nIn file {filepath_errmsg}:\n"
+            "Missing 'END' statement at the end of the 'TFACE' section."
+        )
+
+    return vertices, triangles
+
+
+def _create_tsurf_data(
+    header_name: str,
+    coord_sys_data: dict[str, Any],
+    vertices: list[list[float]],
+    triangles: list[list[int]],
+) -> TSurfData:
```

Comment:
Maybe this should be a class method on TSurfData?

```python
class TSurfData:
    ...

    def from_data(cls, ...) -> Self:
        ...
    # Or, everything
    def from_file(cls, file: FileLike) -> Self:
        ...
```

This pattern is used in other places, and then the function call wraps around it.

I don't think this is necessary to do (at least, now). But just an idea. Maybe for consistency it should be done later, but that will be easier to see after more things exist in xtgeo.io, and we can establish an expected interface for container objects before they become xtgeo objects.